metadata/filter.go

```diff
@@ -0,0 +1,110 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import "strings"
+
+type MetaMap = map[string]string
+
+type metadata struct{}
+
+// Metadata exposes the full implementation
+type Metadata interface {
+	MetaGet
+	MetaCopy
+}
+
+// MetaGet exposes a subset, only Get* methods.
+type MetaGet interface {
+	GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap
+	GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap
+}
+
+// MetaCopy exposes a subset, only Copy* methods.
+type MetaCopy interface {
+	CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string)
+	CopyLabelsWithPrefix(src, dst MetaMap, prefix string)
+}
+
+// NewMetadata is a factory function that return a Metadata interface.
+func NewMetadata() Metadata {
+	return metadata{}
+}
+
+// NewMetaGet is a factory function that return a MetaGet interface.
+func NewMetaGet() MetaGet {
+	return metadata{}
+}
+
+// NewMetaCopy is a factory function that return a MetaCopy interface.
+func NewMetaCopy() MetaCopy {
+	return metadata{}
+}
+
+func (metadata) GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+func (metadata) GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+
+func (metadata) CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string) {
+	copyMetaMapWithPrefix(src, dst, prefix)
+}
+func (metadata) CopyLabelsWithPrefix(src, dst MetaMap, prefix string) {
+	copyMetaMapWithPrefix(src, dst, prefix)
+}
+
+// safeCopy copies a key/value pair to dst.
+// When a key is already present in dst, no copy happens.
+func safeCopy(dst MetaMap, key, val string) {
+	if _, err := dst[key]; !err {
+		dst[key] = val
+	}
+}
+
+// copyMetaMapWithPrefix copies key/value pairs in src adding them to dst.
+// When the prefix is not empty, and src key does not contain the prefix string, no copy happens.
+// When a key in src is already present in dst, no copy happens.
+func copyMetaMapWithPrefix(src, dst MetaMap, prefix string) {
+	if prefix != "" {
+		src := getMetaMapWithPrefix(src, prefix)
+		for key, val := range src {
+			safeCopy(dst, key, val)
+		}
+	} else {
+		for key, val := range src {
+			safeCopy(dst, key, val)
+		}
+	}
+}
+
+// getMetaMapWithPrefix returns a map of key/value pairs contained in src that matches the prefix.
+// When the prefix is not empty, and src key does not contain the prefix string, no copy happens.
+func getMetaMapWithPrefix(src MetaMap, prefix string) MetaMap {
```

Comment:
Just use `map[string]string` and rename to `filterByPrefix`. Create a copy though so the original map remains intact.

Reply:
> Just use map[string]string 
That is fine, but makes the code less concise or aesthetic, but arguably more concrete for lack of symbolic types. 

> and rename to `filterByPrefix`
Sure, we can rename to `filterByPrefix` or anything else. 

> Create a copy though so the original map remains intact.
I believe it does return a copy, because `dst` is a new map, that contains the prefix matches.
Or perhaps I misunderstood? Do you mean create a copy of the `src` map before iterating over it, or maybe passing a copy of `src` into this function ahead of calling the function? Regardless, making a copy of `src` seems needlessly non-optimal, but if there is a concern about corruption then it's probably worth the performance loss. Not a biggy, but I'd like to better understand this one, so if you could please elaborate, that would be great?

Reply:
I've refactored the unexported functions to make an internal copy of the source map, so there is less chance for corruption on the pointer, or anything like that. PTAL